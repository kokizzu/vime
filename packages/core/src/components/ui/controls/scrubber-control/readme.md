# vime-scrubber-control

A control that displays the progression of playback and the amount buffered on a horizontal timeline.
The timeline is a slider (`input[type="range"]`) that can be used to change the current playback time.

If the player is buffering, the scrubber will display an animated candystripe in the porition of the
timeline that has not buffered.

## Visual

<img
  src="https://raw.githubusercontent.com/vime-js/vime/master/packages/core/src/components/ui/controls/scrubber-control/scrubber-control.png"
  alt="Vime scrubber control component"
/>

<!-- Auto Generated Below -->

## Usage

### Angular

```html {7} title="example.html"
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-controls>
      <!-- ... -->
      <vime-scrubber-control></vime-scrubber-control>
    </vime-controls>
  </vime-ui>
</vime-player>
```

### Html

```html {7}
<vime-player>
  <!-- ... -->
  <vime-ui>
    <!-- ... -->
    <vime-controls>
      <!-- ... -->
      <vime-scrubber-control></vime-scrubber-control>
    </vime-controls>
  </vime-ui>
</vime-player>
```

### React

```tsx {6,16}
import React from 'react';
import {
  VimePlayer,
  VimeUi,
  VimeControls,
  VimeScrubberControl,
} from '@vime/react';

function Example() {
  return render(
    <VimePlayer>
      {/* ... */}
      <VimeUi>
        {/* ... */}
        <VimeControls>
          <VimeScrubberControl />
        </VimeControls>
      </VimeUi>
    </VimePlayer>
  );
}
```

### Vue

```html {7,18,26} title="example.vue"
<template>
  <VimePlayer>
    <!-- ... -->
    <VimeUi>
      <!-- ... -->
      <VimeControls>
        <VimeScrubberControl />
      </VimeControls>
    </VimeUi>
  </VimePlayer>
</template>

<script>
  import {
    VimePlayer,
    VimeUi,
    VimeControls,
    VimeScrubberControl,
  } from '@vime/vue';

  export default {
    components: {
      VimePlayer,
      VimeUi,
      VimeControls,
      VimeScrubberControl,
    },
  };
</script>
```

## Properties

| Property          | Attribute           | Description                                                                                                                        | Type      | Default |
| ----------------- | ------------------- | ---------------------------------------------------------------------------------------------------------------------------------- | --------- | ------- |
| `alwaysShowHours` | `always-show-hours` | Whether the timestamp in the tooltip should show the hours unit, even if the time is less than 1 hour (eg: `20:35` -> `00:20:35`). | `boolean` | `false` |
| `hideTooltip`     | `hide-tooltip`      | Whether the tooltip should not be displayed.                                                                                       | `boolean` | `false` |
| `noKeyboard`      | `no-keyboard`       | Prevents seeking forward/backward by using the Left/Right arrow keys.                                                              | `boolean` | `false` |

## CSS Custom Properties

| Name                              | Description                                                                                 |
| --------------------------------- | ------------------------------------------------------------------------------------------- |
| `--scrubber-buffered-bg`          | The background color of the section that indicates how much of the media has been buffered. |
| `--scrubber-loading-stripe-color` | The color of each candystripe displayed when media is buffering.                            |
| `--scrubber-loading-stripe-size`  | The size of each candystripe displayed when media is buffering.                             |

## Dependencies

### Used by

- [vime-default-controls](../default-controls)

### Depends on

- [vime-slider](../../slider)
- [vime-tooltip](../../tooltip)

### Graph

```mermaid
graph TD;
  vime-scrubber-control --> vime-slider
  vime-scrubber-control --> vime-tooltip
  vime-default-controls --> vime-scrubber-control
  style vime-scrubber-control fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
