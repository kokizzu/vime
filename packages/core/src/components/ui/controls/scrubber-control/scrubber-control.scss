vime-scrubber-control {
  /**
   * @prop --scrubber-loading-stripe-size: The size of each candystripe displayed when media is 
   * buffering.
   * @prop --scrubber-loading-stripe-color: The color of each candystripe displayed when media is 
   * buffering.
   * @prop --scrubber-buffered-bg: The background color of the section that indicates how much 
   * of the media has been buffered.
   */

  $scrubber-loading-size: var(--scrubber-loading-stripe-size);
  $scrubber-loading-color: var(--scrubber-loading-stripe-color);
  $scrubber-buffered-bg: var(--scrubber-buffered-bg);

  $slider-track-height: var(--slider-track-height);
  
  // Offset the range thumb in order to be able to calculate the relative progress.
  $scrubber-offset: var(--slider-thumb-width);

  @keyframes progress {
    to {
      background-position: $scrubber-loading-size 0;
    }
  }

  flex: 1;
  cursor: pointer;
  position: relative;
  pointer-events: auto;
  left: calc(#{$scrubber-offset} / 2);
  margin-right: $scrubber-offset;
  margin-bottom: $slider-track-height;
 
  vime-slider,
  progress {
    margin-left: calc(0px - calc(#{$scrubber-offset} / 2));
    margin-right: calc(0px - calc(#{$scrubber-offset} / 2));
    width: calc(100% + #{$scrubber-offset});
    height: $slider-track-height;

    &:hover {
      cursor: pointer;
    }
  }

  vime-slider {
    position: absolute;
    top: 0;
    left: 0;
    z-index: 3;
  }
  
  progress {
    -webkit-appearance: none;
    background: transparent;
    border: 0;
    border-radius: 100px;
    position: absolute;
    left: 0;
    top: 50%;
    padding: 0;
    color: $scrubber-buffered-bg;
    height: $slider-track-height;

    &::-webkit-progress-bar {
      background: transparent;
    }

    &::-webkit-progress-value {
      background: currentColor;
      border-radius: 100px;
      min-width: $slider-track-height;
      transition: width 0.2s ease;
    }

    &::-moz-progress-bar {
      background: currentColor;
      border-radius: 100px;
      min-width: $slider-track-height;
      transition: width 0.2s ease;
    }

    &::-ms-fill {
      border-radius: 100px;
      transition: width 0.2s ease;
    }

    &.loading {
      animation: progress 1s linear infinite;
      background-image: linear-gradient(-45deg,
      $scrubber-loading-color 25%,
      transparent 25%,
      transparent 50%,
      $scrubber-loading-color 50%,
      $scrubber-loading-color 75%,
      transparent 75%,
      transparent);
      background-repeat: repeat-x;
      background-size: $scrubber-loading-size $scrubber-loading-size;
      color: transparent;
      background-color: transparent;
    }
  }
}