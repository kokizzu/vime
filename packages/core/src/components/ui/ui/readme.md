# vime-ui

Simple container that holds a collection of user interface components.

The only important role this component really has is, avoiding overlapping custom UI with the 
native iOS media player UI. Therefore, custom UI is only displayed on iOS if the `playsinline` prop 
is `true`, and the player is not in fullscreen mode.

## Example

```html
<vime-player>
  <!-- ... -->

  <vime-ui>
    <vime-poster />
    <!-- ... -->
  </vime-ui>
</vime-player>
```

<!-- Auto Generated Below -->


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
