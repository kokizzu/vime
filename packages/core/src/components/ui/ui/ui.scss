vime-ui {
  position: relative;
  pointer-events: none;
  z-index: $z-index-ui;
  width: 100%;

  &.hidden {
    display: none;
  }

  &.video {
    position: absolute;
    top: 0;
    left: 0;
    height: 100%;
  }
}