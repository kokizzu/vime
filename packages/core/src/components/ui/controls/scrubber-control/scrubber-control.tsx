import {
  h, Host, Component, Prop, State, Watch, Element, writeTask,
} from '@stencil/core';
import { openPlayerWormhole } from '../../../core/player/PlayerWormhole';
import { PlayerProp, PlayerProps } from '../../../core/player/PlayerProp';
import { formatTime } from '../../../../utils/formatters';
import { createPlayerStateDispatcher, PlayerStateDispatcher } from '../../../core/player/PlayerState';
import { Disposal } from '../../../core/player/Disposal';
import { listen } from '../../../../utils/dom';
import { findRootPlayer } from '../../../core/player/utils';

@Component({
  tag: 'vime-scrubber-control',
  styleUrl: 'scrubber-control.scss',
})
export class ScrubberControl {
  private slider!: HTMLVimeSliderElement;

  private tooltip!: HTMLVimeTooltipElement;

  private dispatch!: PlayerStateDispatcher;

  private keyboardDisposal = new Disposal();

  @Element() el!: HTMLVimeScrubberControlElement;

  @State() timestamp = '';

  @State() endTime = 0;

  /**
   * Whether the timestamp in the tooltip should show the hours unit, even if the time is less than
   * 1 hour (eg: `20:35` -> `00:20:35`).
   */
  @Prop() alwaysShowHours = false;

  /**
   * Whether the tooltip should not be displayed.
   */
  @Prop() hideTooltip = false;

  /**
   * @internal
   */
  @Prop() currentTime: PlayerProps[PlayerProp.CurrentTime] = 0;

  /**
   * @internal
   */
  @Prop() duration: PlayerProps[PlayerProp.Duration] = -1;

  /**
   * Prevents seeking forward/backward by using the Left/Right arrow keys.
   */
  @Prop() noKeyboard = false;

  @Watch('noKeyboard')
  onNoKeyboardChange() {
    this.keyboardDisposal.empty();
    if (this.noKeyboard) return;

    const player = findRootPlayer(this);

    const onKeyDown = (event: KeyboardEvent) => {
      if ((event.key !== 'ArrowLeft') && (event.key !== 'ArrowRight')) return;
      event.preventDefault();
      const isLeftArrow = (event.key === 'ArrowLeft');
      const seekTo = isLeftArrow
        ? Math.max(0, this.currentTime - 5)
        : Math.min(this.duration, this.currentTime + 5);
      this.dispatch(PlayerProp.CurrentTime, seekTo);
    };

    this.keyboardDisposal.add(listen(player, 'keydown', onKeyDown));
  }

  @Watch('duration')
  onDurationChange() {
    // Avoid -1.
    this.endTime = Math.max(0, this.duration);
  }

  /**
   * @internal
   */
  @Prop() buffering: PlayerProps[PlayerProp.Buffering] = false;

  /**
   * @internal
   */
  @Prop() buffered: PlayerProps[PlayerProp.Buffered] = 0;

  /**
   * @internal
   */
  @Prop() i18n: PlayerProps[PlayerProp.I18N] = {};

  connectedCallback() {
    this.dispatch = createPlayerStateDispatcher(this);
    this.timestamp = formatTime(this.currentTime, this.alwaysShowHours);
    this.onNoKeyboardChange();
  }

  disconnectedCallback() {
    this.keyboardDisposal.empty();
  }

  private setTooltipPosition(value: number, percent: number) {
    const tooltipRect = this.tooltip.getBoundingClientRect();
    const scrubberRect = this.el.getBoundingClientRect();
    const bounds = this.el.parentElement!.getBoundingClientRect();
    const leftOffset = scrubberRect.left - bounds.left;
    const rightOffset = bounds.right - scrubberRect.right;
    const leftLimit = bounds.left + leftOffset;
    const rightLimit = bounds.right - rightOffset;
    if (
      (tooltipRect.left + percent > leftLimit)
      && (tooltipRect.right - (100 - percent) < rightLimit)
    ) {
      writeTask(() => {
        this.tooltip.style.left = `${value}px`;
      });
    }
  }

  private onSeek(event: CustomEvent<number>) {
    this.dispatch(PlayerProp.CurrentTime, event.detail);
  }

  private onSeeking(event: MouseEvent) {
    if (this.duration < 0 || this.tooltip.hidden) return;

    if (event.type === 'mouseleave') {
      this.getSliderInput().blur();
      this.tooltip.active = false;
      return;
    }

    const rect = this.el.getBoundingClientRect();
    const percent = Math.max(0, Math.min(100, (100 / rect.width) * (event.pageX - rect.left)));
    this.timestamp = formatTime((this.duration / 100) * percent, this.alwaysShowHours);
    this.setTooltipPosition((percent / 100) * rect.width, percent);

    if (!this.tooltip.active) {
      this.getSliderInput().focus();
      this.tooltip.active = true;
    }
  }

  private getSliderInput() {
    return this.slider.querySelector('input')!;
  }

  render() {
    const sliderValueText = this.i18n.scrubberLabel
      .replace(/{currentTime}/, formatTime(this.currentTime))
      .replace(/{duration}/, formatTime(this.endTime));

    return (
      <Host
        onMouseEnter={this.onSeeking.bind(this)}
        onMouseLeave={this.onSeeking.bind(this)}
        onMouseMove={this.onSeeking.bind(this)}
        onTouchMove={() => { this.getSliderInput().focus(); }}
        onTouchEnd={() => { this.getSliderInput().blur(); }}
      >
        <vime-slider
          step={0.01}
          max={this.endTime}
          value={this.currentTime}
          label={this.i18n.scrubber}
          valueText={sliderValueText}
          onValueChange={this.onSeek.bind(this)}
          ref={(el: any) => { this.slider = el; }}
        />

        <progress
          class={{
            loading: this.buffering,
          }}
          // @ts-ignore
          min={0}
          max={this.endTime}
          value={this.buffered}
          aria-label={this.i18n.buffered}
          aria-valuemin="0"
          aria-valuemax={this.endTime}
          aria-valuenow={this.buffered}
          aria-valuetext={`${((this.endTime > 0) ? (this.buffered / this.endTime) : 0).toFixed(0)}%`}
        >
          % buffered
        </progress>

        <vime-tooltip
          hidden={this.hideTooltip}
          ref={(el: any) => { this.tooltip = el; }}
        >
          {this.timestamp}
        </vime-tooltip>
      </Host>
    );
  }
}

openPlayerWormhole(ScrubberControl, [
  PlayerProp.I18N,
  PlayerProp.CurrentTime,
  PlayerProp.Duration,
  PlayerProp.Buffering,
  PlayerProp.Buffered,
]);
