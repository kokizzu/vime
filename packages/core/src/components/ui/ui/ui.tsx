import {
  h, Host, Component, Prop,
} from '@stencil/core';
import { openPlayerWormhole } from '../../core/player/PlayerWormhole';
import { PlayerProp, PlayerProps } from '../../core/player/PlayerProp';
import { IS_IOS } from '../../../utils/support';

@Component({
  tag: 'vime-ui',
  styleUrl: 'ui.scss',
})
export class UI {
  /**
   * @internal
   */
  @Prop() isVideoView: PlayerProps[PlayerProp.IsVideoView] = false;

  /**
   * @internal
   */
  @Prop() playsinline: PlayerProps[PlayerProp.Playsinline] = false;

  /**
   * @internal
   */
  @Prop() isFullscreenActive: PlayerProps[PlayerProp.IsFullscreenActive] = false;

  render() {
    const canShowCustomUI = !IS_IOS
      || !this.isVideoView
      || (this.playsinline && !this.isFullscreenActive);

    return (
      <Host
        class={{
          hidden: !canShowCustomUI,
          video: this.isVideoView,
        }}
      >
        {canShowCustomUI && <slot />}
      </Host>
    );
  }
}

openPlayerWormhole(UI, [
  PlayerProp.IsVideoView,
  PlayerProp.Playsinline,
  PlayerProp.IsFullscreenActive,
]);
