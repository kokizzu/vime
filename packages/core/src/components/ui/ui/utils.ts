import { getElement } from '@stencil/core';
import { HTMLStencilElement } from '@stencil/core/internal';

export const findUIRoot = (ref: any) => {
  let ui = getElement(ref);

  while (!(/^VIME-UI$/.test(ui?.nodeName))) {
    ui = ui.parentElement as HTMLStencilElement;
  }

  return ui as HTMLVimeUiElement;
};
