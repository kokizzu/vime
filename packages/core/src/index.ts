import { PlayerProp, PlayerProps } from './components/core/player/PlayerProp';
import { PlayerDispatcher, createPlayerDispatcher } from './components/core/player/PlayerDispatcher';
import { withPlayerContext, usePlayerContext } from './components/core/player/PlayerContext';
import { PlayerEvent } from './components/core/player/PlayerEvent';
import { findRootPlayer } from './components/core/player/utils';

export * from './components';

export {
  PlayerProp,
  PlayerProps,
  PlayerEvent,
  findRootPlayer,
  withPlayerContext,
  usePlayerContext,
  PlayerDispatcher,
  createPlayerDispatcher,
};
