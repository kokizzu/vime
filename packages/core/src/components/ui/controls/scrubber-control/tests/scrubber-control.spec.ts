import { SpecPage } from '@stencil/core/testing';
import { newUISpecPage } from '../../../ui/tests';
import { ScrubberControl } from '../scrubber-control';

let page: SpecPage;
let control: HTMLVimeScrubberControlElement;

beforeEach(async () => {
  ({ page } = await newUISpecPage(
    [ScrubberControl],
    '<vime-scrubber-control />',
  ));

  control = page.root!.querySelector('vime-scrubber-control')!;
});

it('should be structurally sound', () => {
  expect(control).toMatchSnapshot();
});
