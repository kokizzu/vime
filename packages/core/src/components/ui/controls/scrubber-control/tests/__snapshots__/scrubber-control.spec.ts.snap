// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be structurally sound 1`] = `
<vime-scrubber-control>
  <vime-slider label="Scrubber" max="0" step="0.01" value="0" valuetext="00:00 of 00:00"></vime-slider>
  <progress aria-label="Buffered" aria-valuemax="0" aria-valuemin="0" aria-valuenow="0" aria-valuetext="0%" max="0" min="0" value="0">
    % buffered
  </progress>
  <vime-tooltip>
    00:00
  </vime-tooltip>
</vime-scrubber-control>
`;
